File: haohaninfo/回測範例/47.py
# -*- coding: UTF-8 -*-
#時間轉數值
def TimetoNumber(time):
 time=time.zfill(8)
 sec=int(time[:2])*360000+int(time[2:4])*6000+int(time[4:6])*100+int(time[6:8])
 return sec

#取I020，依照逗點分隔，並將分隔符號去除
I020 = [ line.strip('\n').split(",") for line in open('Futures_20170815_I020.csv')][1:]

#定義相關變數
MAarray = []
MAValue = 0
STime = TimetoNumber('08450000')
Cycle = 6000
MAlen = 10
#定義上一筆值，提供給策略判斷
lastMAValue=0
lastPrice=0
#倉位為0
index=0


#開始進行MA計算
for i in I020:
 time=i[0]
 price=int(i[4])
 if len(MAarray)==0:
  MAarray+=[price]
 else:
  if TimetoNumber(time)<STime+Cycle:
   MAarray[-1]=price
  else:
   if len(MAarray)==MAlen:
    MAarray=MAarray[1:]+[price]
   else:
    MAarray+=[price]   
   STime = STime+Cycle
 #到達第10分鐘後，開始進行策略判斷
 if len(MAarray)==MAlen:
  MAValue=float(sum(MAarray))/len(MAarray)
  if lastMAValue==0 or lastPrice==0:
   lastMAValue=MAValue
   lastPrice=price
   continue
  if index==0:
   if MAValue<price and lastMAValue>=lastPrice:
    OrderTime=time		#新倉時間紀錄
    OrderPrice=price 	#新倉價格紀錄
    index=1
    print "Buy OrderTime:",OrderTime," OrderPrice:",OrderPrice,
   elif MAValue> price and lastMAValue<=lastPrice:
    OrderTime=time		#新倉時間紀錄
    OrderPrice=price 	#新倉價格紀錄
    index=-1
    print "Sell OrderTime:",OrderTime," OrderPrice:",OrderPrice,
  elif index!=0:
   if index==1:
    if price>=OrderPrice+10 or price<=OrderPrice-10:
     CoverTime=time			#平倉時間紀錄
     CoverPrice=price 		#平倉時間紀錄
     print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",CoverPrice-OrderPrice
     break
    elif i == len(I020)-1:
     CoverTime=time  		#平倉時間紀錄
     CoverPrice=price  		#平倉時間紀錄
     print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",CoverPrice-OrderPrice
   if index==-1:
    if price<=OrderPrice-10 or price>=OrderPrice+10:
     CoverTime=time  			#平倉時間紀錄
     CoverPrice=price 			#平倉時間紀錄
     print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",OrderPrice-CoverPrice
     break
   elif i == len(I020)-1:
     CoverTime=time  		#平倉時間紀錄
     CoverPrice=price  		#平倉時間紀錄
     print " CoverTime:",CoverTime," CoverPrice:",CoverPrice," Profit:",OrderPrice-CoverPrice

File: haohaninfo/交易回測案例/47.R

# 取得分鐘數的函數
getMin <- function(time){
 min <- round(as.numeric(substr(time,nchar(time)-5,nchar(time)))/10000,0)
 return(min)
}

# 載入成交資訊
I020 <- read.csv('Futures_20170815_I020.csv')

A01 <- I020[c(1,5,6)]
names(A01) <- c("INFO_TIME","PRICE","QTY") 
 
# 起始及出場時段
A03 <- subset(A01, INFO_TIME<=as.numeric(13300000)) 

# 初始狀態
Index<-0
MAarray <- numeric(0)
MA <- 0 
lastMA <-0

# 開始進行策略回測
for(i in 2:nrow(A03)){
 
 if(Index==0){
  #動態計算MA
  if(length(MAarray)==0){
   MAarray <- c(A03[i,]$PRICE,MAarray)
  }else if(getMin(A03[i,]$INFO_TIME)==getMin(A03[i-1,]$INFO_TIME)){
   MAarray[1] <- A03[i,]$PRICE
  }else{
   MAarray <- c(A03[i,]$PRICE,MAarray)
   if(length(MAarray)>10){
    MAarray <- MAarray[-11]
   }
  }

  #若尚未滿10 分鐘，則不進行進場判斷
  if(length(MAarray)<10){
   next
  }else if (lastMA==0){
   lastMA = round(sum(MAarray)/10,3)
   next
  }else { 
   MA <- round(sum(MAarray)/10,3)
  }
  #進場判斷
  if(A03[i,]$PRICE>MA && A03[i-1,]$PRICE<lastMA){
    Index=1 
    OrderTime <- A03[i,]$INFO_TIME
        OrderPrice <- A03[i,]$PRICE   
   #cat(Time0,"BUY",OrderPrice,"StopPoint",StopPoint,"\n")
   }else if (A03[i,]$PRICE<MA && A03[i-1,]$PRICE>lastMA){
    Index=-1
    OrderTime <- A03[i,]$INFO_TIME
        OrderPrice <- A03[i,]$PRICE   
       #cat(Time0,"SELL",OrderPrice,"StopPoint",StopPoint,"\n")
   }
   lastMA=MA
  }
  #出場判斷
  if(Index==1){
   if((A03[i,]$PRICE > OrderPrice+5 || A03[i,]$PRICE < OrderPrice-3)){
    CoveryTime <- A03[i,]$INFO_TIME
    CoveryPrice <- A03[i,]$PRICE
    cat("BUY TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",CoveryPrice-OrderPrice,'\n') 
    break
   }else if(i==nrow(A03)){
    CoveryTime <- A03[i,]$INFO_TIME
                  CoveryPrice <- A03[i,]$PRICE
    cat("BUY TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",CoveryPrice-OrderPrice,'\n')  
   }
  }else if(Index==-1){
   if((A03[i,]$PRICE < OrderPrice-5 || A03[i,]$PRICE > OrderPrice+3)){
                Time1 <- A03[i,]$INFO_TIME
                CoveryPrice <- A03[i,]$PRICE
    cat("SELL TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",OrderPrice-CoveryPrice,'\n')
    break
         }else if(i==nrow(A03)){
                Time1 <- A03[i,]$INFO_TIME
                CoveryPrice <- A03[i,]$PRICE
    cat("SELL TIME:",OrderTime,"PRICE:",OrderPrice,"COVERY TIME:",CoveryTime,"C_PRICE:",CoveryPrice,"PROFIT",OrderPrice-CoveryPrice,'\n')
   }
  }  
 }

